### PR TITLE
Check for occupied niches

### PR DESCRIPTION
**Pull request by saethlin**
Implementation of https://github.com/rust-lang/compiler-team/issues/624

Crater run has 62 crates that hit the check, 43 of those are published to crates.io. I see a lot of null function pointers and use of `mem::uninitialized` where the 0x1-filling collides with an enum niche.

But that is with full niche checks; checking transmute, plus any place where that we Copy, Move, or Inspect. Such checking is definitely too thorough to be on by default because it is 2x compile time overhead.

---

During implementation, this ran into https://github.com/llvm/llvm-project/issues/68381

r? @ghost

### CODE REVIEW

**Comment by saethlin** [comment]
This is not done yet. I wish generalized niche checks were as simple as what I whipped up here, and also CI doesn't pass for... reasons.

**Comment by saethlin** [comment]
Wow, I expected worse.

File: compiler/rustc_mir_transform/src/check_niches.rs [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+    fn insert_niche_check(
+        &mut self,
+        block_data: &mut BasicBlockData<'tcx>,
+        new_block: BasicBlock,
+        operand: Operand<'tcx>,
+        niche: NicheKind,
+        source_info: SourceInfo,
+    ) {
+        let type_name =
+            with_no_trimmed_paths!(format!("{:?}", operand.ty(self.local_decls, self.tcx)));
```
Comment (bjorn3) [inline compiler/rustc_mir_transform/src/check_niches.rs:119]: Is there a reason you are not letting the codegen backend do this? That should reduce the crate metadata size by avoiding a string to be stored.

**Comment by saethlin** [comment]
@rust-timer build c5cdc49c3d68ad2a4d68aef9e99ed4e01350deec

**Comment by saethlin** [comment]
@craterbot run mode=build-and-test

**Comment by saethlin** [comment]
Ah yeah I forgot that this adds a niche check on all copy/move/inspect of a Place with a niche. That's probably a bit much. I'll think about that.

**Comment by saethlin** [comment]
I think the problem is the niche checks for `dfa::Result<usize>`. The current codegen for them looks like:
```
    bb26: {
        _66 = move (_68.0: usize);
        _65 = dfa::Result::<usize>::Match(move _66); 
        _432 = _65 as [std::mem::MaybeUninit<u8>; 16] (Transmute);
        _434 = const 0_usize;
        _435 = const 0_usize;
        _433[_434] = _432[_435];
        _436 = const 1_usize;
        _437 = const 1_usize;
        _433[_436] = _432[_437];
        _438 = const 2_usize;
        _439 = const 2_usize;
        _433[_438] = _432[_439];
        _440 = const 3_usize;
        _441 = const 3_usize;
        _433[_440] = _432[_441];
        _442 = const 4_usize;
        _443 = const 4_usize;
        _433[_442] = _432[_443];
        _444 = const 5_usize;
        _445 = const 5_usize;
        _433[_444] = _432[_445];
        _446 = const 6_usize;
        _447 = const 6_usize;
        _433[_446] = _432[_447];
        _448 = const 7_usize;
        _449 = const 7_usize;
        _433[_448] = _432[_449];
        _431 = _433 as u64 (Transmute);
        _450 = Le(_431, const 2_u64);
        _451 = _431 as u128 (IntToInt);
        assert(_450, "occupied niche: {} must be in {}..={} in a {} at offset {} with type {}" _451 const 0_u128 const 2_u128 "dfa::Result<usize>" const 0_u128 "Int(I64, false)") -> [success: bb105, unwind unreachable];
    }
```
That's a lot of MIR.

**Comment by saethlin** [comment]
CI failure on LLVM 15. I think that's the 'ol miscompile.
@bors try @rust-timer queue

**Comment by saethlin** [comment]
It looks like we're emitting a branch hint for all these asserts, which is just a waste when optimizations are off. And also LLVM is eating our checks for bools, for some reason it decides to truncate them. Or something. Maybe our lowering is bad. Anyway, lets try patching both of those:
@bors try @rust-timer queue

**Comment by saethlin** [comment]
I've done a bit more micro-opt, by avoiding ptrtoint for pointer niches, and for partial niches that start at offset 0 I'm just doing a pointer load. It's less LLVM IR at least...
@bors try @rust-timer queue

File: compiler/rustc_mir_transform/src/check_niches.rs (hunk 1 above)
Comment (saethlin) [reply compiler/rustc_mir_transform/src/check_niches.rs]: There isn't any reason. I confirmed that removing this improves compile times, perhaps at some point I'll add this information back.

**Comment by saethlin** [comment]
To reproduce the crash, compile this:
```rust
#[derive(thiserror::Error)]
#[error("...")]
pub enum ErrorEnum {
    Test {
        source: u8, 
    },  
}
```
With a toolchain built with 
```
[rust]
debug = true
codegen-units = 1
```

**Comment by DianQK** [comment]
> To reproduce the crash, compile this:
> 
> ```rust
> #[derive(thiserror::Error)]
> #[error("...")]
> pub enum ErrorEnum {
>     Test {
>         source: u8, 
>     },  
> }
> ```
> 
> With a toolchain built with
> 
> ```
> [rust]
> debug = true
> codegen-units = 1
> ```

I can confirm that this is the same issue as #116976. LLVM 17.0.4 should be released soon.

**Comment by Dylan-DPC** [comment]
@saethlin any updates on this?

**Comment by saethlin** [comment]
Yes; I'm working on an alternative implementation. It's almost ready for a draft PR.

**Comment by saethlin** [comment]
Replaced by https://github.com/rust-lang/rust/pull/121174

**Pull request closed without merging**